nuxeo-mongodb-directory/src/main/java/org/nuxeo/directory/mongodb/MongoDBReference.java

```diff
@@ -0,0 +1,292 @@
+            return getIdsFor(sourceField, sourceId, targetField, session);
+        }
+    }
+
+    /**
+     * Retrieves all target ids associated to the given source id
+     * 
+     * @param sourceId the source id
+     * @param session the mongoDB session
+     * @return the list of target ids
+     * @throws DirectoryException
+     */
+    public List<String> getTargetIdsForSource(String sourceId, MongoDBSession session) throws DirectoryException {
+        return getIdsFor(sourceField, sourceId, targetField, session);
+    }
+
+    @Override
+    public List<String> getSourceIdsForTarget(String targetId) throws DirectoryException {
+        try (MongoDBSession session = getMongoDBSession()) {
+            return getIdsFor(targetField, targetId, sourceField, session);
+        }
+    }
+
+    private List<String> getIdsFor(String queryField, String value, String resultField, MongoDBSession session) {
+        FindIterable<Document> docs = session.getCollection(collection)
+                                             .find(MongoDBSerializationHelper.fieldMapToBson(queryField, value));
+        List<String> ids = StreamSupport.stream(docs.spliterator(), false)
+                                        .map(doc -> doc.getString(resultField))
+                                        .collect(Collectors.toList());
+        return ids;
+    }
+
+    @Override
+    public void setTargetIdsForSource(String sourceId, List<String> targetIds) throws DirectoryException {
+        try (MongoDBSession session = getMongoDBSession()) {
+            setTargetIdsForSource(sourceId, targetIds, session);
+        }
+    }
+
+    /**
+     * Sets all target ids to be associated to the given source id
+     * 
+     * @param sourceId the source id
+     * @param targetIds the target ids
+     * @param session the mongoDB session
+     * @throws DirectoryException
+     */
+    public void setTargetIdsForSource(String sourceId, List<String> targetIds, MongoDBSession session)
+            throws DirectoryException {
+        setIdsFor(sourceField, sourceId, targetField, targetIds, session);
+    }
+
+    @Override
+    public void setSourceIdsForTarget(String targetId, List<String> sourceIds) throws DirectoryException {
+        try (MongoDBSession session = getMongoDBSession()) {
+            setIdsFor(targetField, targetId, sourceField, sourceIds, session);
+        }
+    }
+
+    private void setIdsFor(String field, String value, String fieldToUpdate, List<String> ids, MongoDBSession session) {
+        Set<String> idsToAdd = new HashSet<>();
+        if (ids != null) {
+            idsToAdd.addAll(ids);
+        }
+        List<String> idsToDelete = new ArrayList<>();
+
+        List<String> existingIds = getIdsFor(field, value, fieldToUpdate, session);
+        for (String id : existingIds) {
+            if (idsToAdd.contains(id)) {
+                idsToAdd.remove(id);
+            } else {
+                idsToDelete.add(id);
+            }
```

Comment:
Can be simplified to:

            if (!idsToAdd.remove(id)) {
                idsToDelete.add(id);
            }